### PR TITLE
[CELEBORN-278] Add openStreamWithCredit RPC.

### PR DESCRIPTION
**Pull request by FMX**
### What changes were proposed in this pull request?
Refactor open stream to support buffer stream.

### Why are the changes needed?
Add new rpc to open buffer stream.

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
Integration test.

### CODE REVIEW

File: common/src/main/java/org/apache/celeborn/common/network/protocol/OpenStream.java [hunk 1]
```diff
@@ -24,25 +24,13 @@
...
-  public OpenStream(String shuffleKey, String fileName, int startMapIndex, int endMapIndex) {
-    this(
-        shuffleKey.getBytes(StandardCharsets.UTF_8),
-        fileName.getBytes(StandardCharsets.UTF_8),
-        startMapIndex,
-        endMapIndex);
-  }
-
-  public OpenStream(byte[] shuffleKey, byte[] fileName, int startMapIndex, int endMapIndex) {
+  public OpenStream(byte[] shuffleKey, byte[] fileName) {
```
Comment (waitinfuture) [inline common/src/main/java/org/apache/celeborn/common/network/protocol/OpenStream.java:31]: ```encodedLength``` should also be changed as startMapIndex and endMapIndex are removed

Comment (waitinfuture) [inline common/src/main/java/org/apache/celeborn/common/network/protocol/OpenStream.java:31]: I think refactoring OpenStream will cause incompatibility issue. I think we can try renaming OpenStream to OpenChunkStream, and we should run test case with newer client and older worker to verify.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala [hunk 2]
```diff
@@ -77,25 +78,34 @@ class FetchHandler(val conf: TransportConf) extends BaseMessageHandler with Logg
         rpcSource.updateMessageMetrics(r, 0)
         handleChunkFetchRequest(client, r)
       case r: RpcRequest =>
-        handleOpenStream(client, r)
+        handleOpenStreamLikeRpcs(client, r)
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala:81]: suggest not modify the function name.

**Review by RexXiong — APPROVED** [review]

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala (hunk 2 above)
Comment (FMX) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala:81]: OK.

**Review by RexXiong — APPROVED** [review]

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

File: common/src/main/java/org/apache/celeborn/common/network/protocol/OpenStreamWithCredit.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+import java.nio.charset.StandardCharsets;
+
+import io.netty.buffer.ByteBuf;
+
+/** Buffer stream used in Map partition scenario. */
+public final class OpenStreamWithCredit extends RequestMessage {
+  public final byte[] shuffleKey;
+  public final byte[] fileName;
+  public final int startIndex;
```
Comment (zhongqiangczq) [inline common/src/main/java/org/apache/celeborn/common/network/protocol/OpenStreamWithCredit.java:30]: is it more understanding by replacing member startIndex/endIndex with startReduceIndex/endReduceIndex like OpenStream?

**Pull request merged**